app/services/UIBuilder.scala

```diff
@@ -45,23 +47,21 @@ case object ValueTypeError extends ErrorStrategy {
 class UIBuilder {
   val logger: Logger = Logger(getClass)
 
-  def buildPage(url: String, stanzas: Seq[VisualStanza], errStrategy: ErrorStrategy = NoError)
-               (implicit stanzaIdToUrlMap: Map[String, String], lang: Lang, labels: Labels): Page = {
+  def buildPage(url: String, stanzas: Seq[VisualStanza], errStrategy: ErrorStrategy = NoError)(implicit ctx: UIBuildData): Page = {
     val stanzaTransformPipeline: List[Seq[VisualStanza] => Seq[VisualStanza]] =
       List(expandLabelReferences(Nil), BulletPointBuilder.groupBulletPointInstructions(Nil), Aggregator.aggregateStanzas(Nil), stackStanzas(Nil))
     Page(url, fromStanzas(stanzaTransformPipeline.foldLeft(stanzas){case (s, t) => t(s)}, Nil, errStrategy.default(stanzas)))
   }
 
-  private def expandLabelReferences(acc: List[VisualStanza])(stanzas: Seq[VisualStanza])(implicit labels: Labels, lang: Lang): Seq[VisualStanza] =
+  private def expandLabelReferences(acc: List[VisualStanza])(stanzas: Seq[VisualStanza])(implicit ctx: UIBuildData): Seq[VisualStanza] =
     stanzas match {
       case Nil => acc.reverse
-      case (i: Instruction) +: xs => expandLabelReferences(i.copy(text = TextBuilder.expandLabels(i.text, labels)) :: acc)(xs)
+      case (i: Instruction) +: xs => expandLabelReferences(i.copy(text = TextBuilder.expandLabels(i.text, ctx.labels)) :: acc)(xs)
```

Comment:
I think we need to do this for NoteCallouts as well, because like instructions they can be used to create bullet point lists. I have implemented this is in EG-1373.

Reply:
yes , but its a forward reference to future code ;)